### PR TITLE
fix: Add pagination parameter to canonical URL if present to improve SEO

### PR DESCRIPTION
**Pull request by jonaaix**
### 1. Why is this change necessary?
According to the [Google pagination guidelines](https://developers.google.com/search/docs/specialty/ecommerce/pagination-and-incremental-page-loading#use-urls-correctly), each paginated page should have it's own URL.
> Give each page a unique URL. For example, include a ?page=n query parameter, as URLs in a paginated sequence are treated as separate pages by Google.
Don't use the first page of a paginated sequence as the canonical page. Instead, give each page its own [canonical URL](https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls).

### 2. What does this change do, exactly?
- Add a query parameter on a paginated page to the canonical url.
- Update the canonical url on pagination with the current page number.

### 3. Describe each step to reproduce the issue or behaviour.
1. Open a paginated page, e.g. page 3. 
2. Refresh the page.
3. Open browser inspector and search for the "canonical" meta tag.
4. See the page number is not present in the canonical url.

### 4. Please link to the relevant issues (if any).

https://shopwarecommunity.slack.com/archives/C011VFQT7GB/p1748245912015919

### 5. Checklist

- [1/2] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [-] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

I currently have no development setup and no time to set it up. So I would be glad if someone could test the JS and test implementation. Sorry for that

### CODE REVIEW

File: src/Storefront/Resources/app/storefront/src/plugin/listing/listing-pagination.plugin.js [hunk 1]
```diff
@@ -129,4 +131,16 @@ export default class ListingPaginationPlugin extends FilterBasePlugin {
...
     }
+
+    /**
+     * Update the canonical URL with the new page number.
+     * @param newPageNumber
+     * @private
+     */
+    _updateCanonicalUrl(newPageNumber) {
+        const canonicalMetaTag = document.querySelector('link[rel="canonical"]');
+        const canonicalUrl = new URL(canonicalMetaTag.href);
```
Comment (tobiasberge) [inline src/Storefront/Resources/app/storefront/src/plugin/listing/listing-pagination.plugin.js]: We should add an extra error-handling/check here to verify that `canonicalMetaTag` is found in the DOM and also that the `href` is there. At the moment this can throw an error - which also happens in the jest test: 

https://github.com/shopware/shopware/actions/runs/15283422215/job/42990498187?pr=9936

File: src/Storefront/Resources/app/storefront/test/plugin/listing/listing-pagination.plugin.test.js [hunk 2]
```diff
@@ -93,6 +93,11 @@ describe('listing-pagination.plugin', () => {
         expect(listingPaginationPlugin.getValues).toReturnWith({ 'p': '3' });
         expect(getValuesSpy).toHaveBeenCalledTimes(1);
         expect(changeListingSpy).toHaveBeenCalledTimes(1);
+
+        // Ensure the canonical URL is updated
+        const canonicalMetaTag = document.querySelector('link[rel="canonical"]');
+        const canonicalUrl = new URL(canonicalMetaTag.href);
```
Comment (tobiasberge) [inline src/Storefront/Resources/app/storefront/test/plugin/listing/listing-pagination.plugin.test.js]: Would you also adjust the jest test to support a null case if the `link[rel="canonical"]` element is not found? And for the "normal/asserted" case add  the canonical element to the mock template at the top of the file?

**Review by tobiasberge — CHANGES_REQUESTED** [review]
Hi @btxtiger,

good catch 👍 and thank you for you contribution. Would you mind to add an additional null check in case the canonical element is not found?

File: src/Storefront/Resources/app/storefront/test/plugin/listing/listing-pagination.plugin.test.js (hunk 2 above)
Comment (jonaaix) [reply src/Storefront/Resources/app/storefront/test/plugin/listing/listing-pagination.plugin.test.js]: I hope I understood correctly, fix added

**Review by keulinho — APPROVED** [review]

**Review by BrocksiNet — APPROVED** [review]

**Comment by keulinho** [comment]
@jonaaix ,

thx for your contribution, somehow this thing fall through the cracks, but now it's finally merged 🚀
